### PR TITLE
feat(ENG-818): send additional location params when logging first visit

### PR DESCRIPTION
**Pull request by carlmw**
Depends on https://github.com/oaknational/oak-consent-api/pull/22

### CODE REVIEW

File: src/core/client.ts [hunk 1]
```diff
@@ -303,9 +303,30 @@ export class OakConsentClient implements ConsentClient {
...
     } catch (error) {
       this.onError(error);
     }
   };
+
+  /**
+   * Extracts additional context from the current page
+   * to be logged with the user's visit
+   */
+  private extractLocationParams() {
```
Comment (simonrose121) [inline src/core/client.ts:320]: Should this have a return type of `LogUserAdditionalParams`? I'm thinking it'd be clearer with a type, it'd also mean updating the `LogUserAdditionalParams` to have optional properties and dropping the `Partial` from the `logUser` method. We're also always returning the `url`, so that's not actually optional?

**Review by simonrose121 — CHANGES_REQUESTED** [review]
Looks good and well tested, just a minor comment about types.

File: src/core/client.ts (hunk 1 above)
Comment (mantagen) [reply src/core/client.ts:320]: Good idea @simonrose121

File: src/core/client.ts [hunk 2]
```diff
@@ -303,9 +303,30 @@ export class OakConsentClient implements ConsentClient {
...
+    const url = new URL(window.location.href);
+    return {
+      utmSource: url.searchParams.get("utm_source") ?? undefined,
```
Comment (mantagen) [inline src/core/client.ts:323]: Edge case but if utm_source etc is present but without value, this will be set as empty string rather than undefined here. I think that's desired, but just noted that behaviour differs for referrerUrl

**Review by mantagen — APPROVED** [review]
Looks good. Agree with @simonrose121  might make sense to have the more flexible type

File: src/core/client.ts (hunk 1 above)
Comment (carlmw) [reply src/core/client.ts:320]: @simonrose121 brace yourself, you're going to get my opinions 😬

TL;DR I'm ok with making this change if it is a convention within the org, but I disagree with sharing the type and the need for it on a private method. 

1. As a personal preference I try to rely on inference for local/private functions that are by their nature implementation detail — I could have left the body of this function inside `logFirstVisitByUser` and everything would be fine. Extracting it was an exercise in improving readability. If this were public 💯 there would be a contract to maintain with a type (especially with the complexity of the object being returned).
2. The only slight downside is that the type warning will be a few lines above this method (but I'd argue it is worth that for maintenance cost)
3. The `LogUserAdditionalParams` type is intended to convey the contract between `Network` and the API itself, where all of these params are optional. While the `url` is required in this instance, If I were to make `url` required in the type, that would impose that restriction on any other caller of `Network#logUser`. That wouldn't be my intention (ok it's unlikely that anything else would ever use that method but you never know!).
4. Further to ☝🏻 I should really make `additionalParams` an optional argument on `Network#logUser` which I will do
5. Thinking about it further I should probably be more explicit in the tests that the `url` is required instead of hiding it amongst and example with all params. I'll add another spec! 
6. I'm also a bit wary of exporting tons of types (until they become truly universal), it makes it tough to eliminate them should they become redundant.

File: src/core/client.ts (hunk 2 above)
Comment (carlmw) [reply src/core/client.ts:323]: Good catch! I think it is entirely likely that a bad tool/copy-paste job will leave empty values behind

File: src/core/client.ts (hunk 1 above)
Comment (mantagen) [reply src/core/client.ts:320]: Once upon a time i tried to find consensus on how we [organised types on OWA](https://www.notion.so/oaknationalacademy/How-to-organising-shared-types-in-OWA-878e83b7fcf14e67b033b715e2759ffa), but got zero engagement.

Having read your explanation and gone back to look at the code, I completely agree that this method should _not_ use the `LogUserAdditionalParams` type. Moreover, if it were me I'd probably have a `LocationParams` type, and possilbe the `LogUserAdditionalParams` could extend that type.

I'm more than happy for you to either proceed go with your own suggestions above!

Comment (simonrose121) [reply src/core/client.ts:320]: @carlmw No problem! 

I'd actually missed that it was a private function, so I agree we shouldn't be using (or exporting) the `LogUserAdditionalParams` type. The purpose of that type being as a contract between the `Network` and API makes sense.

I think @mantagen has a good point above, but it's not a blocker for me if you feel it unnecessary.

**Review by simonrose121 — APPROVED** [review]
Happy to approve based on the above conversation.

File: src/core/client.ts (hunk 1 above)
Comment (carlmw) [reply src/core/client.ts:320]: @mantagen it's very hard to get people to have opinions (or even share them when they do have them 😉). In the past i've forgone the big document in favour of discussions just like this. Happening more organically around specific examples. Easier to then extract them into some sort of standard. Still people might not feel too strongly about stuff 🤷🏻‍♂️  

Interested to understand why you'd break this type up before there is a second example needing `LocationParams` (assuming that type would consist of `url` and `referrerUrl`)

I could see a benefit to splitting the type for hinting purposes. One for the location params and another for the UTMs which could be unioned to give 👇🏻 

Screenshot 2024-07-09 at 15 38 01

Comment (mantagen) [reply src/core/client.ts:320]: Agreed.

It wasn't breaking the type up, more just defining it the return type.

I'm not sure exactly where I stand on it, however I'm drawn to the argument as follows:

```ts
function extractLocationParams() {
  // to understand what is extracted would involve scanning code to be executed, possibly additional context
}
```
whereas:
```ts
type LocationParams = {
  // very clear what is extracted
}
function extractLocationParams(): LocationParams {
  // 
}
```

In addition to the clarity point, I think it encourages intentionality when amending the function.
